### PR TITLE
counters: implement `Count` for `Option`/`Result`

### PR DESCRIPTION
**Pull request by hawkw**
This commit adds implementations of the `counters::Count` trait for
`Option<T> where T: Count` and `Result<T, E> where T: Count, E: Count`.
Currently, we aren't using these anywhere, but we _will_ be! :D

### CODE REVIEW

File: lib/counters/src/lib.rs [hunk 1]
```diff
@@ -77,3 +78,132 @@ macro_rules! count {
...
+    }
+}
+
+impl Count for core::convert::Infallible {
+    type Counters = AtomicU32;
+    #[allow(clippy::declare_interior_mutable_const)]
+    const NEW_COUNTERS: Self::Counters = AtomicU32::new(0);
+
+    fn count(&self, _counters: &Self::Counters) {
+        // `Infallible`s are not made. They should NEVER be made. We
```
Comment (cbiffle) [inline lib/counters/src/lib.rs:182]: I agree with the Litany Against Infallible comment here, but would suggest replacing the code in this function with the "discharge an uninhabited enum" pattern:

`match self {}`

That ensures that there's no potential panic site generated or unreachable code.

File: lib/counters/src/lib.rs [hunk 2]
```diff
@@ -77,3 +78,132 @@ macro_rules! count {
...
+        #[cfg(not(debug_assertions))]
+        armv6m_atomic_hack::AtomicU32Ext::fetch_add(
+            _counters,
+            1,
+            Ordering::Relaxed,
+        );
+    }
+}
+
+impl Count for () {
```
Comment (cbiffle) [inline lib/counters/src/lib.rs:197]: This impl is kind of the first time where I've been a little concerned about us overflowing a U32. How often are we counting `()`?

File: lib/counters/src/lib.rs (hunk 1 above)
Comment (hawkw) [reply lib/counters/src/lib.rs:182]: ah, good call, thank you!

File: lib/counters/src/lib.rs (hunk 2 above)
Comment (hawkw) [reply lib/counters/src/lib.rs:197]: This impl is^H^Hwill be used for counting the `Ok(())` arm of a `Result<(), T> where T: Count`. In particular, for fallible IPC operations, Idol will generate an enum variant like 
```rust
enum ${name_of_ipc_interface}Event {
    ${name_of_fallible_ipc_op}(#[count(children)] Result<(), #{the_fallible_op_error_type}>,
    // ... other variants ...
}
```

so, there's no _global_ counter of `()`s; it's just used to instantiate a counter for the `Ok` side of a `Result` counter in a particular enum variant.

We could use similar counters in hand-written code; IIRC we have some counted ringbuffers that contain variants that are `Result<(), Error>` that aren't currently counting the `Ok`/`Err` variants separately, but could be. 

Does that explanation make sense?

Comment (cbiffle) [reply lib/counters/src/lib.rs:188]: It does, thanks.

**Pull request merged**